Add steps for upgrading external database (#1614)
theforeman/foreman-documentation: guides/common/modules/proc_upgrading-the-external-database.adoc

[id="Upgrading_the_External_Database_{context}"]
= Upgrading the External Database

You can upgrade an external database from {EL} 7 to {EL} 8 while upgrading {Project} from {ProjectVersionPrevious} to {ProjectVersion}.

.Prerequisites

> * Create a new {EL} 8 based host for PostgreSQL server that follows the external database on {EL} 8 documentation.

Author: @evgeni Is there a section for using external databases or am I missing something?

> * Create a new {EL} 8 based host for PostgreSQL server that follows the external database on {EL} 8 documentation.

Reviewer: I was expecting "follows the external database documentation" to link to https://docs.theforeman.org/nightly/Installing_Server/index-katello.html#using-external-databases_foreman, yes.

> * Create a new {EL} 8 based host for PostgreSQL server that follows the external database on {EL} 8 documentation.

Author: Thanks! It's in the Installation Guide so I have to use a different link than xref.

For more information, see https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/deploying_different_types_of_servers/using-databases#migrating-to-a-rhel-8-version-of-postgresql_using-postgresql[Migrating to a RHEL 8 version of PostgreSQL].

.Procedure
. Create a backup.
. If you keep the same hostname on the new PostgreSQL server, update the DNS record for the host to prevent the need to do updates on {Project} itself.

> . If the new PostgreSQL server has a new host name or a different CNAME, update the reference to the external database on {Project}:

Reviewer: If you perform any of these steps *before* restoring the DB on the new system, you will see errors, which will confuse the reader (especially if you use the Installer method and the installer will fail).

Reviewer: I'd move them towards the end, see also the wording suggestion in https://github.com/theforeman/foreman-documentation/pull/1614#discussion_r1021241684

Author: There really wasn't a place to move these since there weren't many steps in the Procedure to begin with but it has been reworded.

+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# {foreman-installer} \

> --foreman-db-host newpostgres.example.com \
> --katello-candlepin-db-host newpostgres.example.com \
> --foreman-proxy-content-pulpcore-postgresql-host newpostgres.example.com

Reviewer: ```suggestion
--foreman-db-host _newpostgres.example.com_ \
--katello-candlepin-db-host _newpostgres.example.com_ \
--foreman-proxy-content-pulpcore-postgresql-host _newpostgres.example.com_
```

italics to indicate things that need replacement

[id="Upgrading_the_External_Database_{context}"]
= Upgrading the External Database

You can upgrade an external database from {EL} 7 to {EL} 8 while upgrading {Project} from {ProjectVersionPrevious} to {ProjectVersion}.

.Prerequisites
* Create a new {EL} 8 based host for PostgreSQL server that follows the external database on {EL} 8 documentation.
For more information, see https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/deploying_different_types_of_servers/using-databases#migrating-to-a-rhel-8-version-of-postgresql_using-postgresql[Migrating to a RHEL 8 version of PostgreSQL].

.Procedure

> . On a virtual machine, take a snapshot.
> . On a physical machine, create a backup.

Reviewer 2: This should be reduced to:
```suggestion
. Create a backup.
```

This is because the procedure later states "Copy the backup database".

Additionally, many sysadmins will keep a virtual server around for a while as a backup, instead of separate backups/snapshots. In most virtualization platforms you can just shut it down anyway.

. If you keep the same hostname on the new PostgreSQL server, update the DNS record for the host to prevent the need to do updates on {Project} itself.

> . If the new PostgreSQL server has a new host name or a different CNAME, update the reference to the external database on {Project}:

Reviewer 2: I think this is confusing. Let's first start at the requirement. The requirement is that the Foreman server can reach the DB server.

There are multiple ways to achieve this.

First of all, a new hostname is used and the procedure below uses the installer to change the configs accordingly.

Then the other is that the same hostname is used. To make sure it reaches the right server, you can use two techniques.

You described changing the DNS record, but this has some potential issues. Any DNS record has a TTL (Time To Live), which allows any DNS server to cache it. If you don't plan for it and you have a high TTL value (86400 seconds, or 1 day, is common) then it can take a long time for the migration to propagate. A common technique is to plan for it, lower the TTL in advance (at least $TTL) and then change it. Once it's stable it can be raised again.

However during the migration it may still resolve to the old host if you have multiple caching servers. That's why many admins actually prefer to move the IP to the new server instead. So you use some temporary IP to completely provision & prepare the server and then add the old IP while removing it from the old server (or shutting that down).

Some also combine that and do migrate to a new IP but add both IPs to the same server during the migration so the stale DNS cache isn't an issue.

In my mind those are all fairly standard operating procedures so I wonder how much of it should be written out.

Author: @evgeni related to my comment above regarding this point. Your thoughts?

Reviewer: I think I'd slightly reword this, making it less ambiguous:
* create a backup 
* restore the backup on the new server
* If {Project} can reach the new database server via the old name, no further changes are required. Otherwise reconfigure {Project} to use the new name: <installer invocation>

All further details are left as an exercise to the reader, as we do not know how their setup looks like.

Reviewer 2: > create a backup

Should we recommend an offline backup? So stop all services and then create a backup to prevent data loss?

Reviewer: Given https://github.com/theforeman/foreman-documentation/blob/a9d7313f01d8fbe0d7a4a88db72583a24d1f7385/guides/common/assembly_using-external-databases.adoc#L10-L12
I would tend to say "nah, not our pint"

(but we probably should add a similar sentence to non-Satellite builds of the documentation)

[id="Upgrading_the_External_Database_{context}"]
= Upgrading the External Database

You can upgrade an external database from {EL} 7 to {EL} 8 while upgrading {Project} from {ProjectVersionPrevious} to {ProjectVersion}.

.Prerequisites
* Create a new {EL} 8 based host for PostgreSQL server that follows the external database on {EL} 8 documentation.

> ifdef::satellite[]
> For more information, see https://access.redhat.com/documentation/en-us/red_hat_satellite/{ProjectVersion}/html/installing_satellite_server_in_a_connected_network_environment/performing-additional-configuration#using-external-databases_satellite[Using External Databases with {Project}].
> endif::[]
> ifdef::katello[]
> For more information, see https://docs.theforeman.org/{ProjectVersion}/Installing_Server/index-katello.html#using-external-databases_foreman[Using External Databases with {Project}].
> endif::[]

Reviewer: This can be written using macros in one go:
```suggestion
ifdef::katello,orcharhino,satellite[]
For more information, see {InstallingServerDocURL}using-external-databases_{project-context}[Using External Databases with {Project}].
endif::[]
```

(we do not include external DB for foreman builds, no idea why

Reviewer 2: > (we do not include external DB for foreman builds, no idea why

8710bde2c7185c057d7f9a71313d7c166dae5fda. When I was reading through it, it only worked if you had Katello. If we add more conditionals to hide the Katello-specifics (like candlepin & pulp's databases) it can be included.

.Procedure
. Create a backup.
. Restore the backup on the new server.
. If {Project} reaches the new database server via the old name, no further changes are required.
Otherwise reconfigure {Project} to use the new name:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# {foreman-installer} \
--foreman-db-host _newpostgres.example.com_ \
--katello-candlepin-db-host _newpostgres.example.com_ \
--foreman-proxy-content-pulpcore-postgresql-host _newpostgres.example.com_
----

> . Copy the backup database to the new host.
> . Restore the backup database on the new host.

Reviewer: These steps are already done above
```suggestion
```
